testing/txsim/client.go

```diff
@@ -0,0 +1,277 @@
+		}
+
+		// set the latest height
+		if status.SyncInfo.EarliestBlockHeight > height {
+			height = status.SyncInfo.EarliestBlockHeight
+		}
+	}
+	return &TxClient{
+		rpcClients:  rpcClients,
+		encCfg:      encCfg,
+		chainID:     chainID,
+		pollTime:    pollTime,
+		height:      height,
+		lastUpdated: time.Now(),
+	}, nil
+}
+
+func (tc *TxClient) Tx() sdkclient.TxBuilder {
+	builder := tc.encCfg.TxConfig.NewTxBuilder()
+	return builder
+}
+
+func (tc *TxClient) ChainID() string {
+	return tc.chainID
+}
+
+func (tc *TxClient) Height() int64 {
+	tc.mtx.Lock()
+	defer tc.mtx.Unlock()
+	return tc.height
+}
+
+func (tc *TxClient) updateHeight(newHeight int64) int64 {
+	tc.mtx.Lock()
+	defer tc.mtx.Unlock()
+	if newHeight > tc.height {
+		tc.height = newHeight
+		tc.lastUpdated = time.Now()
+		return newHeight
+	}
+	return tc.height
+}
+
+func (tc *TxClient) LastUpdated() time.Time {
+	tc.mtx.Lock()
+	defer tc.mtx.Unlock()
+	return tc.lastUpdated
+}
+
+// WaitForNBlocks uses WaitForHeight to wait for the given number of blocks to
+// be produced.
+func (tc *TxClient) WaitForNBlocks(ctx context.Context, blocks int64) error {
+	return tc.WaitForHeight(ctx, tc.Height()+blocks)
+}
+
+// WaitForHeight continually polls the network for the latest height. It is
+// concurrently safe.
+func (tc *TxClient) WaitForHeight(ctx context.Context, height int64) error {
+	// check if we can immediately return
+	if height <= tc.Height() {
+		return nil
+	}
+
+	ticker := time.NewTicker(tc.pollTime)
+	for {
+		select {
+		case <-ticker.C:
+			// check if we've reached the target height
+			if height <= tc.Height() {
+				return nil
+			}
+			// check when the last time we polled to avoid concurrent processes
+			// from polling the network too often
+			if time.Since(tc.LastUpdated()) < tc.pollTime {
+				continue
+			}
+
+			// ping a node for their latest height
+			status, err := tc.Client().Status(ctx)
+			if err != nil {
+				return fmt.Errorf("error getting status from rpc server: %w", err)
+			}
+
+			latestHeight := tc.updateHeight(status.SyncInfo.LatestBlockHeight)
+			// check if the new latest height is greater or equal than the target height
+			if latestHeight >= height {
+				return nil
+			}
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (tc *TxClient) WaitForTx(ctx context.Context, txID []byte) (*coretypes.ResultTx, error) {
+	for i := 0; i < maxRetries; i++ {
+		resp, err := tc.Client().Tx(ctx, txID, false)
+		if err != nil {
+			// tx still no longer exists
+			if strings.Contains(err.Error(), "not found") {
+				time.Sleep(tc.pollTime)
+				continue
+			}
+			return nil, err
+		}
+
+		if resp.TxResult.Code != 0 {
+			return nil, fmt.Errorf("non zero code delivering tx (%d): %s", resp.TxResult.Code, resp.TxResult.Log)
+		}
+
+		return resp, nil
+	}
+	return nil, errTimedOutWaitingForTx
+}
+
+// Client multiplexes the RPC clients
+func (tc *TxClient) Client() *http.HTTP {
+	tc.mtx.Lock()
+	defer tc.mtx.Unlock()
```

Comment:
definitely not blocking, but perhaps if we use read mutexes here, we can be more consistent when mutex usage when modifying local state, such as in the next method